cmd/template/cluster/provider/common.go

```diff
@@ -51,3 +64,80 @@ func newCAPIV1Alpha3ClusterCR(config ClusterCRsConfig, infrastructureRef *corev1
 
 	return cluster
 }
+
+func runMutation(ctx context.Context, client k8sclient.Interface, input []byte) ([]byte, error) {
+	obj := &unstructured.Unstructured{}
+	dec := apiyaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+	_, gvk, err := dec.Decode(input, nil, obj)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	// Create a DiscoveryClient to get the GVR.
+	dc, err := discovery.NewDiscoveryClientForConfig(client.RESTConfig())
+	if err != nil {
+		return []byte{}, err
+	}
+
+	// Mapping needed for the dynamic client.
+	mapping, err := findGVR(gvk, dc)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	// Create a DynamicClient to execute our request.
+	dyn, err := dynamic.NewForConfig(client.RESTConfig())
+	if err != nil {
+		return []byte{}, err
+	}
+
+	namespace, namespaced, err := unstructured.NestedString(obj.Object, "metadata", "namespace")
+	if err != nil {
+		return []byte{}, err
+	}
+
+	var defaultedObj *unstructured.Unstructured
+	// Execute our request as a `DryRun` - no resources will be persisted.
+	if namespaced {
+		defaultedObj, err = dyn.Resource(mapping.Resource).Namespace(namespace).Create(ctx, obj, metav1.CreateOptions{DryRun: []string{"All"}})
+		if err != nil {
+			// TODO handle different kinds of errors (e.g. validation) here.
+			return []byte{}, err
+		}
+	} else {
+		defaultedObj, err = dyn.Resource(mapping.Resource).Create(ctx, obj, metav1.CreateOptions{DryRun: []string{"All"}})
+		if err != nil {
+			// TODO handle different kinds of errors (e.g. validation) here.
+			return []byte{}, err
+		}
+	}
+
+	// Strip `managedFields` for better readability.
+	unstructured.RemoveNestedField(defaultedObj.Object, "metadata", "managedFields")
+	// Strip some metadata fields for better UX.
+	unstructured.RemoveNestedField(defaultedObj.Object, "metadata", "creationTimestamp")
+	unstructured.RemoveNestedField(defaultedObj.Object, "metadata", "generation")
+	unstructured.RemoveNestedField(defaultedObj.Object, "metadata", "selfLink")
+	unstructured.RemoveNestedField(defaultedObj.Object, "metadata", "uid")
+
+	// Unstructured to JSON.
+	buf := new(bytes.Buffer)
+	err = dec.Encode(defaultedObj, buf)
```

Comment:
Can you not write it to YAML directly with `yaml.Marshal`?

Reply:
I am not sure because the object is `unstructured.Unstructured` and from the examples I was able to find they seem to rely on the `apimachinery` serializer. I can try later on though.

Reply:
It's just a few extra lines, so not worth spending too much time on. I was just curious.